oauth2client/flask.py

```diff
@@ -0,0 +1,299 @@
+        try:
+            credentials = flow.step2_exchange(request.args['code'])
+        except FlowExchangeError as e:
+            current_app.logger.exception(e)
+            return e.message, 400
+
+        # Save the credentials to the session.
+        session['oauth2credentials'] = credentials.to_json()
+
+        # Get the user's profile from Google and save it to the session.
+        if 'profile' in self.scopes:
+            session['oauth2profile'] = self._request_user_info()
+
+        if self.authorize_callback:
+            self.authorize_callback()
+
+        return redirect(session.pop('oauth2return', '/'))
+
+    @property
+    def credentials(self):
+        ctx = _app_ctx_stack.top
+
+        if not hasattr(ctx, 'oauth2_credentials'):
+            serialized = session.get('oauth2credentials')
+
+            if not serialized:
+                return None
+
+            ctx.oauth2_credentials = Credentials.new_from_json(serialized)
+
+        return ctx.oauth2_credentials
+
+    def has_credentials(self):
+        return self.credentials and not self.credentials.invalid
+
+    @property
+    def email(self):
+        """Returns the user's email address.
+
+        The email address is provided by the current credentials' id_token. The
+        use can change their email, it's more appropriate to use user_id for
+        a unique, user-specific identifier.
+        """
+        if not self.credentials:
+            return None
+        return self.credentials.id_token['email']
+
+    @property
+    def user_id(self):
+        """Returns the user's email address.
+
+        The id is provided by the current credentials' id_token.
+        """
+        if not self.credentials:
+            return None
+        return self.credentials.id_token['sub']
+
+    @property
+    def profile(self):
+        """Return's the user's full Google profile.
+
+        This is only available if "profile" is specified in the scopes.
+        """
+        return session.get('oauth2profile', None)
+
+    def required(self, f):
+        """Decorator to require OAuth2 credentials for a view.
+        """
+        @wraps(f)
+        def decr(*args, **kwargs):
+            if not self.has_credentials():
+                return redirect(
+                    url_for('oauth2.authorize', return_url=request.url))
+            else:
+                return f(*args, **kwargs)
+        return decr
+
+    def http(self, *args, **kwargs):
+        """Returns an authorized http instance.
+
+        Can only be called if there are valid credentials for the user, such
+        as inside of a view that is decorated with @required.
+
+        Args:
+            *args: Positional arguments passed to httplib2.Http constructor.
+            **kwargs: Positional arguments passed to httplib2.Http constructor.
+        """
+        if not self.credentials:
+            raise NoCredentialsError()
+        return self.credentials.authorize(httplib2.Http(*args, **kwargs))
```

Comment:
The http instance? I don't think so, and they aren't thread safe anyway.

The credentials are cached in the flask context.